notification/src/handler/notification/notification.handler.js

```diff
@@ -377,12 +377,23 @@ async function getTransactionTypeAndStateOrNull(notificationRequestItem) {
   )
   const adyenEventCode = notificationRequestItem.eventCode
   const adyenEventSuccess = notificationRequestItem.success
+  const adyenEventFailReason = notificationRequestItem.reason
 
   // eslint-disable-next-line max-len
-  const adyenEvent = _.find(
-    adyenEvents,
-    (e) => e.eventCode === adyenEventCode && e.success === adyenEventSuccess,
-  )
+  let adyenEvent = null;
+  _.find(adyenEvents, (e) => {
+    if (e.eventCode === adyenEventCode && e.success === adyenEventSuccess) {
+      if (adyenEventFailReason && e.reason === adyenEventFailReason) {
+        adyenEvent = e;
+        return true;
+      }
+      if (!adyenEvent) {
+        adyenEvent = e;
+      }
+    }
+    return false;
+  });
```

Comment:
how about to change it to smth like:
`const adyenEvent = adyenEvents
    .filter((e) => e.eventCode === adyenEventCode && e.success === adyenEventSuccess)
    .filter((e) => (e.reason && e.reason === adyenEventReason) || !e.reason);`

so it will return events that match by reason first, if they don't or no reason is available it will return the one w/o/ the reason (original pending event).

Reply:
i think it would be better to do it in one iteration. also it would be easier if we have to add more conditions to this part in the future.

Reply:
it's just this code is harder to read when there are multiple ifs, assignments, and returns when in fact we don't care what _.find returns as we don't assign its result to anything. What I suggested looks cleaner IMO, happy to ask for a 3rd opinion.

Reply:
i asked it to chatgpt and selected your way 😄 updated it and slightly changed this part: 
`find((e) => (e.reason && e.reason === adyenEventReason) || (!e.reason && !adyenEventReason));`
i added `&& !adyenEventReason` part because otherwise fallback return true in most of the cases even though it shouldn't